### PR TITLE
Toggl labs/close on start

### PR DESCRIPTION
**Pull request by almirbi**
## :star2: What does this PR do?

* block tabs on start time entry
* add block page
* add timer to block page

## :bug: Recommendations for testing

Go to Settings -> Distraction blocker, make sure to add a URL, like https://www.youtube.com/
Open that URL, like https://www.youtube.com/, in a new tab
Start time entry for anywhere in the extension

Should show block page, still not synced though

### CODE REVIEW

File: src/scripts/website-blocking/WebsiteBlocking.ts [hunk 1]
```diff
@@ -3,7 +3,11 @@ import browser from 'webextension-polyfill';
 import WebsiteBlockingApi from "./WebsiteBlockingApi";
 import WebsiteBlockingConverter from "./WebsiteBlockingConverter";
 
-const db = browser.extension.getBackgroundPage().db;
+const db = () => browser.extension.getBackgroundPage().db;
```
Comment (EternallLight) [inline src/scripts/website-blocking/WebsiteBlocking.ts:6]: what's the difference? it isn't fully initialised here?

File: src/scripts/website-blocking/WebsiteBlocking.ts [hunk 2]
```diff
@@ -3,7 +3,11 @@ import browser from 'webextension-polyfill';
...
+
+function youShallNotPass(tabId) {
```
Comment (EternallLight) [inline src/scripts/website-blocking/WebsiteBlocking.ts:8]: let's move this to yet another static function
this class is going to be shrunk down anyway

File: src/scripts/website-blocking/WebsiteBlocking.ts [hunk 3]
```diff
@@ -117,8 +121,33 @@ class WebsiteBlocking {
...
-}
 
-export default WebsiteBlocking;
+  static async closeOnStart() {
+    
+    const websiteBlockingList = JSON.parse(await db().get('websiteBlockingList') || '[]');
 
+    const blockedSites = websiteBlockingList.map(entry => entry.url);
 
+    console.log(blockedSites)
```
Comment (EternallLight) [inline src/scripts/website-blocking/WebsiteBlocking.ts:131]: I forget to remove those often, too 😀

File: src/scripts/website-blocking/WebsiteBlocking.ts [hunk 4]
```diff
@@ -117,8 +121,33 @@ class WebsiteBlocking {
...
+    if (blockedSites.length > 0) {
+      const blockedTabs = await browser.tabs.query({ url: blockedSites });
+      blockedTabs.forEach(tab => youShallNotPass(tab.id));
+    }
+  };
+
+  static async blockSite(tabId, changeInfo) {
+    const url = changeInfo.pendingUrl || changeInfo.url;
+
```
Comment (EternallLight) [inline src/scripts/website-blocking/WebsiteBlocking.ts:140]: Let's put early return here if `url` is falsy.
This function seems to be triggered a lot - e.g. when title is changed on a tab.

**Review by EternallLight — CHANGES_REQUESTED** [review]

**Comment by EternallLight** [comment]
And stop button doesn't seem to work for me
image
image

File: src/scripts/website-blocking/WebsiteBlocking.ts (hunk 1 above)
Comment (almirbi) [reply src/scripts/website-blocking/WebsiteBlocking.ts]: I think it's too early when I use it in backgorund.js, it was failing.

**Review by EternallLight — APPROVED** [review]

**Pull request merged**